File: code/python/02-structured-output/few_shot_comparison.py
"""Zero-shot vs few-shot classification comparison.

Demonstrates the reliability gain from few-shot examples for sentiment
classification. Prints both results side-by-side with token counts so
you can measure the reliability/cost trade-off directly.

See docs/01-foundations/02-prompt-engineering.md — "Few-Shot Prompting"

When few-shot is worth the extra tokens:
  - Classification where the output format MUST be a single word/label.
  - Tasks where the model often adds unwanted explanation.
  - Edge cases where zero-shot returns inconsistent casing or phrasing.
For anything more complex, switch to structured output (json_schema).
"""

from __future__ import annotations

import os

import tiktoken
from openai import OpenAI

MODEL = "gpt-4o"
ALLOWED_LABELS = {"Positive", "Negative", "Neutral"}

ZERO_SHOT_SYSTEM = (
    "Classify the sentiment of the following text as exactly one of: "
    "Positive, Negative, or Neutral."
)

FEW_SHOT_SYSTEM = """\
Classify the sentiment of the following text as exactly one of: \
Positive, Negative, or Neutral.
Respond with exactly one word.

Examples:
Text: "I love this product!" → Positive
Text: "This is absolutely terrible." → Negative
Text: "It arrived on time." → Neutral
"""

# Tricky input: mildly positive but hedged — models often disagree zero-shot.
TEST_INPUT = (
    "The new update is fine, I guess. Not bad, but nothing to get excited about."
)


def _count(messages: list[dict]) -> int:
    enc = tiktoken.encoding_for_model(MODEL)
    total = (
        sum(3 + sum(len(enc.encode(v)) for v in msg.values()) for msg in messages) + 3
    )
    return total


def classify(system_prompt: str, text: str, client: OpenAI) -> tuple[str, int, int]:
    """Return (classification_label, estimated_prompt_tokens, actual_prompt_tokens)."""
    messages = [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": f'Text: "{text}"'},
    ]
    estimated_tokens = _count(messages)
    response = client.chat.completions.create(
        model=MODEL,
        messages=messages,
        temperature=0,
        max_tokens=10,
    )
    actual_prompt_tokens = response.usage.prompt_tokens
    return response.choices[0].message.content.strip(), estimated_tokens, actual_prompt_tokens


def format_status(label: str) -> str:
    """Return whether the model followed the requested one-word label format."""
    return "ok" if label in ALLOWED_LABELS else "drift"


def main() -> None:
    client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])

    print(f'Input: "{TEST_INPUT}"\n')
    print(
        f"{'Approach':<12} {'Result':<12} {'Format':<8} {'Est. prompt':>12} {'API prompt':>12}"
    )
    print("-" * 64)

    zero_result, zero_estimated, zero_actual = classify(
        ZERO_SHOT_SYSTEM, TEST_INPUT, client
    )
    print(
        f"{'Zero-shot':<12} {zero_result:<12} {format_status(zero_result):<8} "
        f"{zero_estimated:>12} {zero_actual:>12}"
    )

    few_result, few_estimated, few_actual = classify(FEW_SHOT_SYSTEM, TEST_INPUT, client)
    print(
        f"{'Few-shot':<12} {few_result:<12} {format_status(few_result):<8} "
        f"{few_estimated:>12} {few_actual:>12}"
    )

    overhead = few_estimated - zero_estimated
    print(f"\nFew-shot overhead: +{overhead} estimated prompt tokens per request")
    print("A drift status means the model ignored the exact one-word label format.")


if __name__ == "__main__":
    main()

File: code/nodejs/02-structured-output/few_shot_comparison.ts
/**
 * Zero-shot vs few-shot classification comparison.
 *
 * Demonstrates reliability gain from few-shot examples for sentiment
 * classification. Prints results side-by-side with token counts.
 * See: docs/01-foundations/02-prompt-engineering.md
 */

import OpenAI from "openai";
import { encoding_for_model } from "js-tiktoken";

const MODEL = "gpt-4o";
const ALLOWED_LABELS = new Set(["Positive", "Negative", "Neutral"]);

const ZERO_SHOT_SYSTEM =
  "Classify the sentiment of the following text as exactly one of: " +
  "Positive, Negative, or Neutral.";

const FEW_SHOT_SYSTEM = `Classify the sentiment of the following text as exactly one of: Positive, Negative, or Neutral.
Respond with exactly one word.

Examples:
Text: "I love this product!" → Positive
Text: "This is absolutely terrible." → Negative
Text: "It arrived on time." → Neutral
`;

const TEST_INPUT =
  "The new update is fine, I guess. Not bad, but nothing to get excited about.";

function countTokens(messages: OpenAI.Chat.ChatCompletionMessageParam[]): number {
  const enc = encoding_for_model("gpt-4o");
  let total = 3; // reply primer
  for (const msg of messages) {
    total += 3;
    for (const val of Object.values(msg)) {
      if (typeof val === "string") total += enc.encode(val).length;
    }
  }
  enc.free();
  return total;
}

async function classify(
  systemPrompt: string,
  text: string,
  client: OpenAI
): Promise<{ label: string; estimatedPromptTokens: number; actualPromptTokens: number }> {
  const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
    { role: "system", content: systemPrompt },
    { role: "user", content: `Text: "${text}"` },
  ];
  const estimatedPromptTokens = countTokens(messages);
  const response = await client.chat.completions.create({
    model: MODEL,
    messages,
    temperature: 0,
    max_tokens: 10,
  });
  return {
    label: response.choices[0].message.content?.trim() ?? "",
    estimatedPromptTokens,
    actualPromptTokens: response.usage?.prompt_tokens ?? 0,
  };
}

function formatStatus(label: string): string {
  return ALLOWED_LABELS.has(label) ? "ok" : "drift";
}

async function main(): Promise<void> {
  const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

  console.log(`Input: "${TEST_INPUT}"\n`);
  console.log(
    `${"Approach".padEnd(12)} ${"Result".padEnd(12)} ${"Format".padEnd(8)} ${"Est. prompt".padStart(12)} ${"API prompt".padStart(12)}`
  );
  console.log("-".repeat(64));

  const zero = await classify(ZERO_SHOT_SYSTEM, TEST_INPUT, client);
  console.log(
    `${"Zero-shot".padEnd(12)} ${zero.label.padEnd(12)} ${formatStatus(zero.label).padEnd(8)} ${String(zero.estimatedPromptTokens).padStart(12)} ${String(zero.actualPromptTokens).padStart(12)}`
  );

  const few = await classify(FEW_SHOT_SYSTEM, TEST_INPUT, client);
  console.log(
    `${"Few-shot".padEnd(12)} ${few.label.padEnd(12)} ${formatStatus(few.label).padEnd(8)} ${String(few.estimatedPromptTokens).padStart(12)} ${String(few.actualPromptTokens).padStart(12)}`
  );

  const overhead = few.estimatedPromptTokens - zero.estimatedPromptTokens;
  console.log(`\nFew-shot overhead: +${overhead} estimated prompt tokens per request`);
  console.log("A drift status means the model ignored the exact one-word label format.");
}

main().catch(console.error);
